crates/sui-core/src/authority/submitted_transaction_cache.rs

```diff
@@ -0,0 +1,343 @@
+        submitter_client_addr: Option<IpAddr>,
+    ) {
+        let mut inner = self.inner.write();
+
+        // We allow 1 submission for the initial submission, and then the retry tolerance + amplification factor
+        // for subsequent submissions.
+        let max_allowed_submissions = 1 + self.retry_tolerance + amplification_factor;
+
+        if let Some(metadata) = inner.transactions.get(digest) {
+            if metadata.submitted_round < submitted_round {
+                let old_round = metadata.submitted_round;
+                debug!(
+                    "Transaction {digest} re-submitted at round {submitted_round} (previously at round {old_round})",
+                );
+
+                if let Some(txns) = inner.transactions_by_round.get_mut(&old_round) {
+                    txns.remove(digest);
+                    if txns.is_empty() {
+                        inner.transactions_by_round.remove(&old_round);
+                    }
+                }
+
+                inner
+                    .transactions_by_round
+                    .entry(submitted_round)
+                    .or_default()
+                    .insert(*digest);
+
+                let metadata = inner.transactions.get_mut(digest).unwrap();
+                metadata.submitted_round = submitted_round;
+            }
+        } else {
+            // First time we're submitting this transaction, however we will wait till
+            // we see the transaction in consensus output to increment the submission count.
+            let metadata = SubmissionMetadata {
+                submitted_round,
+                submission_count: 0,
+                max_allowed_submissions,
+                submitter_client_addr,
+            };
+
+            inner.transactions.insert(*digest, metadata);
+            inner
+                .transactions_by_round
+                .entry(submitted_round)
+                .or_default()
+                .insert(*digest);
+
+            debug!(
+                "First submission of transaction {digest} at round {submitted_round} (max_allowed: {max_allowed_submissions})",
+            );
+        }
+    }
+
+    /// Increments the submission count when we see a transaction in consensus output.
+    /// This tracks how many times the transaction has appeared in consensus (from any validator).
+    /// Returns the spam weight and submitter client address if the transaction exceeds allowed submissions.
+    pub(crate) fn increment_submission_count(
+        &self,
+        digest: &TransactionDigest,
+    ) -> Option<(Weight, Option<IpAddr>)> {
+        let mut inner = self.inner.write();
+
+        if let Some(metadata) = inner.transactions.get_mut(digest) {
+            metadata.submission_count += 1;
+
+            if metadata.submission_count > metadata.max_allowed_submissions {
+                // TODO(fastpath): Reevaluate spam weight calculation. For simplicity, we use a fixed spam weight of 1 for now.
```

Comment:
I think this mechanism is good. We just need to point out that every time `increment_submission_count()` is called and `max_allowed_submissions` is exceeded, a spam weight is added.